ouroboros-consensus/src/Ouroboros/Consensus/NodeNetwork.hs

```diff
@@ -49,148 +52,184 @@ import           Ouroboros.Consensus.ChainSyncServer
 import           Ouroboros.Consensus.Ledger.Abstract
 import           Ouroboros.Consensus.Util.Condense
 import           Ouroboros.Consensus.Util.Orphans ()
-import           Ouroboros.Consensus.Util.ThreadRegistry
 
 import qualified Ouroboros.Storage.ChainDB.API as ChainDB
 
 import           Ouroboros.Consensus.Node
 
 
-{-------------------------------------------------------------------------------
-  Network layer
--------------------------------------------------------------------------------}
-
--- TODO something nicer than eCS, eBF, bytesCS, and bytesBF. Mux them over one
--- channel.
-data NetworkProvides m up blk = NetworkProvides {
-      -- | Notify network layer of new upstream node
-      --
-      -- NOTE: Eventually it will be the responsibility of the network layer
-      -- itself to register and deregister peers.
-      addUpstream   :: forall eCS eBF bytesCS bytesBF.
-                       (Exception eCS, Exception eBF)
-                    => up
-                    -> NodeComms m (ChainSync (Header blk) (Point blk)) eCS bytesCS
-                       -- Communication for the Chain Sync protocol
-                    -> NodeComms m (BlockFetch blk)                     eBF bytesBF
-                       -- Communication for the Block Fetch protocol
-                    -> m ()
-
-      -- | Notify network layer of a new downstream node
-      --
-      -- NOTE: Eventually it will be the responsibility of the network layer
-      -- itself to register and deregister peers.p
-    , addDownstream :: forall eCS eBF bytesCS bytesBF.
-                       (Exception eCS, Exception eBF)
-                    => NodeComms m (ChainSync (Header blk) (Point blk)) eCS bytesCS
-                       -- Communication for the Chain Sync protocol
-                    -> NodeComms m (BlockFetch blk)                     eBF bytesBF
-                       -- Communication for the Block Fetch protocol
-                    -> m ()
-    }
+data SharedState m s a = SharedState (m s) (s -> a)
```

Comment:
Just like @avieth, I'm not a fan of this `SharedState`. I realise why it is needed (I added the `clientRegistered` `TMVar` stuff in the first place to appease the invariants in the block fetch client), but it would be nicer if we could solve it in another way.

Reply:
BTW, why is the `Semigroup` instance needed again? After removing it, the demo and the consensus tests still compile.

Reply:
Just thinking out loud:

Does `SharedState` have to be so general? In particular, does it have to parameterise over `s`? Typically, I like to be as general as possible too, but in this case, being more specific might be clearer?

For example, can we rename it to `InitNetwork(Apps) m a` and fix `s` to `TMVar m ()`?

Alternatively, could it maybe take the `TMVar` as an argument?

Reply:
I probably wasn't clear; `Semigroup` instance would be useful to create multiple versions.  In the demo we use `SharedState`'s functor instance to map `SharedState (TMVar m s) (NetworkApplication ...)` into `SharedState (TMVar m s) (Versions ....)`. `Versions` is basically a `Map versionNumber (NetworkApplication ...)` (it's slightly more general than that but for this context it's enough to think of it as that). Both sides will agree on a single version to run.  To build a non singleton we need a way to combine multiple `SharedState m s (Versions ...)` into one, that's how semigroup instance could be used.  Right now, we have just a single protocol version so the semigroup instance is not really needed.

Reply:
I will rename it and I'll fix the `s` variable.